### PR TITLE
fix external load balancer for HA and upgrading process

### LINKED ISSUE #136: External load balancer and HA upgrade issue
Current, the definition of the load balancer defined in `cluster.yaml` does not take into account when provision an HA cluster.
Therefore there will be a glitch during the upgrade process because the first master will be temporarily unavailable.

### PR DESCRIPTION
**Pull request by chanwit**
This PR implements a feature for setting up an HA cluster with an external load balancer.

Fixes #136

### CODE REVIEW

File: pkg/plan/resource/kubeadm_init.go [hunk 1]
```diff
@@ -229,17 +230,20 @@ func (ki *KubeadmInit) deserializeSecret(fileName, namespace string) (*corev1.Se
...
-	return buildKubeadmInitPlan(remotePath, strings.Join(ki.IgnorePreflightErrors, ","), ki.UseIPTables, &ignored).Undo(
-		runner, plan.EmptyState)
+	return buildKubeadmInitPlan(
+		remotePath,
+		strings.Join(ki.IgnorePreflightErrors, ","),
+		ki.UseIPTables, ki.ControlPlaneEndpoint, &ignored).Undo(runner, plan.EmptyState)
 }
 
-func buildKubeadmInitPlan(path string, ignorePreflightErrors string, useIPTables bool, output *string) plan.Resource {
+func buildKubeadmInitPlan(path string, ignorePreflightErrors string, useIPTables bool, controlPlaneEndpoint string, output *string) plan.Resource {
```
Comment (jrryjcksn) [inline pkg/plan/resource/kubeadm_init.go]: It looks like `controlPlaneEndpoint` isn't used in this function...

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 2]
```diff
@@ -164,11 +165,25 @@ func (a *MachineActuator) create(ctx context.Context, cluster *clusterv1.Cluster
 // don't miss any updates. The plan is derived from the original seed node plan and stored in a config map
 // for use by the actuator.
 func (a *MachineActuator) initializeMasterPlanIfNecessary(installer *os.OS) error {
-	master, err := a.getMasterNode() // Only one can exist at this point
+
+	// we also use this method to mark the first master as the "originalMaster"
```
Comment (jrryjcksn) [inline pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: This is a _much_ nicer way to track original master 👍

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 3]
```diff
@@ -164,11 +165,25 @@ func (a *MachineActuator) create(ctx context.Context, cluster *clusterv1.Cluster
...
 	}
-	if master.Annotations[planKey] == "" {
+	if _, exist := originalMasterNode.Labels[originalMasterLabel]; !exist {
+		if err := a.setNodeLabel(originalMasterNode, originalMasterLabel, ""); err != nil {
+			return err
+		}
+	}
+
+	/*
+		master, err := a.getMasterNode() // Only one can exist at this point
```
Comment (jrryjcksn) [inline pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: Can we remove this comment?

File: pkg/plan/resource/kubeadm_join.go [hunk 4]
```diff
@@ -33,6 +33,8 @@ type KubeadmJoin struct {
 	CertificateKey string `structs:"certificateKey" plan:"hide"`
 	// IgnorePreflightErrors is optionally used to skip kubeadm's preflight checks.
 	IgnorePreflightErrors []string `structs:"ignorePreflightErrors"`
+	// External Load Balancer IP address to be used instead of masterIP
+	ExternalLoadBalancer string `structs:"externalLoadBalancer" plan:"hide"`
```
Comment (jrryjcksn) [inline pkg/plan/resource/kubeadm_join.go]: I'm curious as to why this is hidden in the plan. One side effect of hiding any attribute is that we won't be able to determine that it has changed when we go to update a node. If we decide at some later time to swap out the load balancer after first standing up a new one, the machine_actuator wouldn't know to reconstruct the node with the new reference (thinking it hadn't changed)...

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go (hunk 2 above)
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: thank you!

File: pkg/plan/resource/kubeadm_join.go (hunk 4 above)
Comment (chanwit) [reply pkg/plan/resource/kubeadm_join.go]: I don't have any particular reason.
I just saw that the `masterIP` field is hidden, so I did the same to this field.

I'll remove the plan:hide from this field.

**Comment by chanwit** [comment]
> According to the docs, we only need to choose the first control-plane master. If that node is replaced with a fresh master in the future for whatever real-life reason, and it doesn't have the label, upgrades would break with the current logic as it wouldn't find that master.

@saada Method `getOriginalMaster` works exactly in that way.

1. if it found a node with the label, that node is the original master
2. if there's no labelled node, the first node of that list becomes the original master

**Comment by chanwit** [comment]
@saada I'll make this mechanism more robust by applying the singleton pattern to `getOriginalMaster`. Let me think what's the best way to implement it.

**Comment by saada** [comment]
Thanks for your response. I misunderstood the logic. Approved 👍

**Comment by saada** [comment]
@chanwit , there's a bug in wksctl that we should fix where the kubeadm configmap that we generate uses a mix of private ips and public ips... 

Specifically this block.

```yaml
apiServer:
  certSANs:
  - localhost
  - 3.80.254.97 # public
  - 10.99.239.66 # private (original master)
  - 18.206.98.91 # public
```

Can we fix this bug in this branch as well as it could be a source of many other issues? It's also very inconsistent. All control plane traffic should go through private ips by default.

Let me know your thoughts on this.

**Comment by palemtnrider** [comment]
> @chanwit , there's a bug in wksctl that we should fix where the kubeadm configmap that we generate uses a mix of private ips and public ips...
> 
> Specifically this block.
> 
> ```yaml
> apiServer:
>   certSANs:
>   - localhost
>   - 3.80.254.97 # public
>   - 10.99.239.66 # private (original master)
>   - 18.206.98.91 # public
> ```
> 
> Can we fix this bug in this branch as well as it could be a source of many other issues? It's also very inconsistent. All control plane traffic should go through private ips by default.
> 
> Let me know your thoughts on this.

Hi @saada please open a separate ticket for this.  I'd like to get this PR merged and the more we pile in the longer it will take.

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 5]
```diff
@@ -479,9 +499,14 @@ func (a *MachineActuator) update(ctx context.Context, cluster *clusterv1.Cluster
 		}
 		return a.kubeadmUpOrDowngrade(machine, node, installer, version, planKey, planJSON, worker)
 	}
-	if err = a.performActualUpdate(installer, machine, node, nodePlan, c); err != nil {
-		return err
-	}
+	//
+	// OK not upgrade or downgrade? so why this?
+	// This is non-logical update becuse it's already done by kubeadm?
```
Comment (jrryjcksn) [inline pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: This was intended to support updates such as turning a working into a master, etc. that doesn't fall under the "upgrade" heading.

File: pkg/plan/resource/kubeadm_init.go [hunk 6]
```diff
@@ -253,7 +257,8 @@ func buildKubeadmInitPlan(path string, ignorePreflightErrors string, useIPTables
 		// certificates of the primary control plane in the kubeadm-certs
 		// Secret, and prints the value for --certificate-key to STDOUT.
 		&Run{Script: plan.ParamString(
-			withoutProxy("kubeadm init --config=%s --ignore-preflight-errors=%s --experimental-upload-certs"), &path, &ignorePreflightErrors),
+			withoutProxy("kubeadm init --config=%s --ignore-preflight-errors=%s --experimental-upload-certs"),
```
Comment (palemtnrider) [inline pkg/plan/resource/kubeadm_init.go]: From the documentation I only see --experimental-upload-certs available in 1.14.  Have you tested on other versions?

File: pkg/plan/resource/kubeadm_join.go [hunk 7]
```diff
@@ -33,6 +33,8 @@ type KubeadmJoin struct {
 	CertificateKey string `structs:"certificateKey" plan:"hide"`
 	// IgnorePreflightErrors is optionally used to skip kubeadm's preflight checks.
 	IgnorePreflightErrors []string `structs:"ignorePreflightErrors"`
+	// External Load Balancer IP address to be used instead of masterIP
```
Comment (palemtnrider) [inline pkg/plan/resource/kubeadm_join.go]: Isn't strictly an IP.

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 8]
```diff
@@ -49,12 +50,13 @@ import (
 )
 
 const (
-	planKey          string = "wkp.weave.works/node-plan"
-	masterLabel      string = "node-role.kubernetes.io/master"
-	controllerName   string = "wks-controller"
-	controllerSecret string = "wks-controller-secrets"
-	bootstrapTokenID string = "bootstrapTokenID"
-	clusterName      string = "firekube"
+	planKey             string = "wkp.weave.works/node-plan"
```
[the remark this answers is not in the document]
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: acked

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 9]
```diff
@@ -49,12 +50,13 @@ import (
...
+	masterLabel         string = "node-role.kubernetes.io/master"
+	originalMasterLabel string = "wkp.weave.works/original-master"
```
[the remark this answers is not in the document]
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: like `wks-firekube` ?

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go (hunk 5 above)
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: reverted

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 10]
```diff
@@ -758,33 +798,21 @@ func (a *MachineActuator) checkIfOriginalMasterNotAtVersion(kubernetesVersion st
 }
 
 func (a *MachineActuator) getOriginalMasterNode() (*corev1.Node, error) {
```
[the remark this answers is not in the document]
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: nop, we shouldn't do that.
there might have a small gap that the seed node has not been labeled yet, but we need to check for the original master.

this logic is from @jrryjcksn AFAIC

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 11]
```diff
@@ -164,11 +166,19 @@ func (a *MachineActuator) create(ctx context.Context, cluster *clusterv1.Cluster
...
 func (a *MachineActuator) initializeMasterPlanIfNecessary(installer *os.OS) error {
-	master, err := a.getMasterNode() // Only one can exist at this point
+
+	// we also use this method to mark the first master as the "originalMaster"
+	originalMasterNode, err := a.getOriginalMasterNode()
 	if err != nil {
 		return err
 	}
-	if master.Annotations[planKey] == "" {
+	if _, exist := originalMasterNode.Labels[originalMasterLabel]; !exist {
```
[the remark this answers is not in the document]
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: The assumption made by the previous code was somehow inconsistent.
It randomly returns a master from a list of masters.

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 12]
```diff
@@ -419,10 +430,19 @@ func (a *MachineActuator) update(ctx context.Context, cluster *clusterv1.Cluster
...
+	if currentPlan == planJSON {
 		contextLog.Info("Machine and node have matching plans; nothing to do")
 		return nil
 	}
+
+	if diffedPlan, err := plandiff.GetUnifiedDiff(currentPlan, planJSON); err == nil {
+		contextLog.Info("........................ DIFF PLAN ........................")
+		contextLog.Info(diffedPlan)
+	} else {
+		contextLog.Infof("DIFF PLAN Error: %v", err)
```
[the remark this answers is not in the document]
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: acked

File: pkg/apis/wksprovider/controller/wksctl/machine_actuator.go [hunk 13]
```diff
@@ -758,33 +798,21 @@ func (a *MachineActuator) checkIfOriginalMasterNotAtVersion(kubernetesVersion st
...
-			return nil, gerrors.Wrapf(err, "failed to retrieve internal node address")
-		}
-		if internalAddress == addr {
+		_, isOriginalMaster := node.Labels[originalMasterLabel]
+		if isOriginalMaster {
 			return node, nil
 		}
 	}
-	return nil, errors.New("No original master node found")
+	// Hack for H/A testing
```
[the remark this answers is not in the document]
Comment (chanwit) [reply pkg/apis/wksprovider/controller/wksctl/machine_actuator.go]: it seems I rebased @jrryjcksn codes here.

File: pkg/plan/resource/kubeadm_init.go (hunk 6 above)
Comment (chanwit) [reply pkg/plan/resource/kubeadm_init.go]: we currently allow to start the cluster with only k8s 1.14.1.
I'll fix this in another PR.

File: pkg/plan/resource/kubeadm_join.go (hunk 7 above)
Comment (chanwit) [reply pkg/plan/resource/kubeadm_join.go]: acked

**Comment by palemtnrider** [comment]
Rebased on master.  Will merge when tests pass.

**Pull request merged**